hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java

```diff
@@ -392,6 +397,24 @@ private static void initializeUsedBytesAndBlockCount(DatanodeStore store,
     kvData.setBlockCount(blockCount);
   }
 
+  /**
+   * A container is empty if:
+   * - The container is closed
+   * - There are no blocks in its block table.
+   *
+   * Empty containers are eligible for deletion.
+   */
+  public static boolean isEmpty(DatanodeStore store,
+      KeyValueContainerData container) throws IOException {
+    if (container.isOpen()) {
+      return false;
```

Comment:
Why the container is considered to be empty when it is not open?  It is common that the container has blocks when it transitions into states like  CLOSING.

Reply:
When container is open it is considered as not empty. When in any other state(other than open state) we are checking in block table to see if there are any block present and then we decide container is empty or not.

Reply:
Yes keeping block count metadata can go wrong sometime which leads to wrong interpretation of empty container. 
Now we check if any real block exists in container(not counting all block count), Which is done in isEmpty() method.